SPIFlashLogger.device.lib.nut

```diff
@@ -0,0 +1,847 @@
+        _sectorsMap = [];
+
+        local hasWrongSector = false;
+
+        // Read all the metadata
+        local found_last_pos = false;
+        for (local sector = 0; sector < _sectors; sector++) {
+            local sectorItem =
+                LoggerSector(_start + sector * SPIFLASHLOGGER_SECTOR_SIZE,
+                    _start + (sector + 1) * SPIFLASHLOGGER_SECTOR_SIZE,
+                    _flash,
+                    this);
+            sectorItem.init();
+            _sectorsMap.push(sectorItem);
+            //
+            // Buffer is circular therefore we are not expecting
+            // free sector before latest working sector
+            // Note: throw error in case of several start codes
+            if (sectorItem.isStartWriting() && !sectorItem.isStoppedWriting()) {
+                // By default writing address
+                // is rounded to the chunk
+                _at_pos = sectorItem.getWritePosition();
+                _at_sec = sector;
+                // Check that it is a new start position
+                if (found_last_pos)
+                    // Found several start positions.
+                    hasWrongSector = true;
+                found_last_pos = true;
+            } else if (sectorItem.isFree() && !found_last_pos) {
+                found_last_pos = true;
+                _at_pos = sectorItem.getWritePosition();
+                _at_sec = sector;
+            }
+        }
+        // Clear flash if it has wrong sectors
+        if (hasWrongSector) {
+            eraseAll(true);
+        }
+    }
+
+    //
+    // Erase sectors from start to end
+    //
+    // @param {int} - from sector address
+    // @param {int} - till sector address
+    // @param {bool} - skip preparing sector
+    function _erase(start_sector = null, end_sector = null, preparing = false) {
+        if (start_sector == null) {
+            start_sector = 0;
+            end_sector = _sectors;
+        }
+
+        if (end_sector == null)
+            end_sector = start_sector + 1;
+
+        if (start_sector < 0 || end_sector > _sectors)
+            throw "Invalid format request";
+
+        _enable();
+
+        for (local sector = start_sector; sector < end_sector; sector++) {
+            // Erase the requested sector
+            _flash.erasesector(_start + (sector * SPIFLASHLOGGER_SECTOR_SIZE));
+            // reinit current sector as free
+            if (!preparing)
+                _sectorsMap[sector].init();
+        }
+        _disable();
+    }
+}
+
+class SPIFlashLogger.LoggerSector {
+    // position for reading
+    _pos_read = 0;
+    // position for writing
+    _pos_write = 0;
+    // start/end address
+    _start = 0;
+    _end = 0;
+    // spi flash object
+    _flash = null;
+    _logger = null;
+    // indicates if we need to erase sector before write
+    _eraseBeforeWrite = false;
+
+    // Default status of sector is free
+    // which is equal to the 0xFFFFFF
+    static SECTOR_STATUS_FREE = 0;
```

Comment:
These should be moved to constants

Reply:
ok, thanks, it is a replacement for a flags above